src/vm/eventpipe.cpp

```diff
@@ -38,6 +40,110 @@ extern "C" void InitProvidersAndEvents();
 extern "C" void InitProvidersAndEvents();
 #endif
 
+EventPipeEventPayload::EventPipeEventPayload(byte *pData, unsigned int length)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    m_pData = pData;
+    m_pBlobs = NULL;
+    m_blobCount = 0;
+    m_performedAllocation = false;
+
+    m_size = length;
+}
+
+EventPipeEventPayload::EventPipeEventPayload(EventData **pBlobs, unsigned int blobCount)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    m_pData = NULL;
+    m_pBlobs = pBlobs;
+    m_blobCount = blobCount;
+    m_performedAllocation = false;
+
+    S_UINT32 tmp_size = S_UINT32(blobCount) * S_UINT32(sizeof(EventData));
+    if (tmp_size.IsOverflow()){
+        m_pBlobs = NULL;
+        m_blobCount = 0; // ?
+        m_size = 0; // ?
+    }
+    else{
+        m_size = tmp_size.Value();
+    }
+}
+
+EventPipeEventPayload::~EventPipeEventPayload()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    if(m_performedAllocation){
+        delete[] m_pData;
+        m_pData = NULL;
+    }
+}
+
+void EventPipeEventPayload::Flatten()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    if(m_size > 0){
+        if (!this->IsFlattened()){
+            m_pData = new BYTE[m_size];
+            m_performedAllocation = true;
+            this->CopyData(m_pData);
+        }
+    }
+}
+
+void EventPipeEventPayload::CopyData(BYTE *pDst)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    if(m_size > 0){
+        if(this->IsFlattened()){
+            memcpy(pDst, m_pData, m_size);
+        }
+
+        else if(m_pBlobs != NULL){
+            unsigned int offset = 0;
+            for(int i=0; i<m_blobCount; i++){
+                memcpy(pDst + offset, m_pBlobs, sizeof(EventData));
```

Comment:
Addressed in https://github.com/dotnet/coreclr/pull/13347/commits/851943c5d5dc653dc1625a376c26bc2d6cf6a587